akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala

```diff
@@ -612,94 +713,57 @@ private[akka] class Shard(
-
-                } else {
-                  entityIdsWaitingForWrite match {
-                    case OptionVal.None =>
-                      // No actor running and no write in progress, start actor and deliver message when started
-                      // Note; we only do this if remembering, otherwise the buffer is an overhead
-                      if (VerboseDebug)
-                        log.debug("Buffering message [{}] to [{}] and starting actor", payload.getClass, id)
-                      appendToMessageBuffer(id, msg, snd)
-                      waitForAsyncWrite(id, RememberEntitiesShardStore.AddEntities(Set(id)))(sendMsgBuffer)
-
-                    case OptionVal.Some(ids) if ids.contains(id) =>
-                      // No actor running and write in progress for this particular id, buffer message for deliver when
-                      // write completes
-                      if (VerboseDebug)
-                        log.debug(
-                          "Buffering message [{}] to [{}] because of write in progress for it",
-                          payload.getClass,
-                          id)
-                      appendToMessageBuffer(id, msg, snd)
-
-                    case OptionVal.Some(otherIds) =>
-                      // No actor running and write in progress for some other entity id, stash message for deliver when
-                      // unstash happens on async write complete
-                      if (VerboseDebug)
-                        log.debug(
-                          "Stashing message [{}] to [{}] because of write in progress for [{}]",
-                          payload.getClass,
-                          id,
-                          otherIds)
-                      stash()
-                  }
-                }
-            }
+          entities.entityState(id) match {
+            case Some(Remembering) =>
+              appendToMessageBuffer(id, msg, snd)
+            case Some(Passivating(_)) =>
+              appendToMessageBuffer(id, msg, snd)
+            case Some(Active(ref)) =>
+              if (VerboseDebug) log.debug("Delivering message of type [{}] to [{}]", payload.getClass, id)
+              touchLastMessageTimestamp(id)
+              ref.tell(payload, snd)
+            case Some(RememberedButNotCreated) | Some(WaitingForRestart) =>
+              if (VerboseDebug)
+                log.debug(
+                  "Delivering message of type [{}] to [{}] (starting because known but not running)",
+                  payload.getClass,
+                  id)
+
+              val actor = getOrCreateEntity(id)
+              touchLastMessageTimestamp(id)
+              actor.tell(payload, snd)
+            case None if !entityIdsWaitingForWrite.isDefined =>
+              // No actor running and no write in progress, start actor and deliver message when started
+              // Note; we only do this if remembering, otherwise the buffer is an overhead
```

Comment:
note about buffer but then we always buffer below? (could be my fault since earlier)

Reply:
It looks like a lie!

Reply:
And a lie in master